src/e2e/app.ts

```diff
@@ -1139,6 +1145,42 @@ export class App {
     await actions.click();
   }
 
+  async simulateTab(): Promise<void> {
+    const keyboard = (await this.page).keyboard;
+    await keyboard.press("Tab");
+  }
+
+  async getActiveElement(): Promise<puppeteer.ElementHandle<Element>> {
+    return (await this.page).evaluateHandle<ElementHandle>(
+      () => document.activeElement
+    );
+  }
+
+  async getElementByRoleAndLabel(
+    role: string,
+    name: string
+  ): Promise<puppeteer.ElementHandle<Element>> {
+    return (await this.document()).findByRole(role, {
+      name,
+    });
+  }
+
+  async getElementByQuerySelector(
+    query: string
+  ): Promise<puppeteer.ElementHandle<Element>> {
+    return (await this.page).evaluateHandle<ElementHandle>(
+      (query) => document.querySelector(query),
+      query
+    );
+  }
+
+  async compareElementHandles(
+    e1: puppeteer.ElementHandle<Element>,
+    e2: puppeteer.ElementHandle<Element>
+  ): Promise<boolean> {
+    return (await this.page).evaluate((e1, e2) => e1 === e2, e1, e2);
+  }
```

Comment:
If we expose all this then it's quite likely the e2e tests will deteriorate into a big mess with lots of details of the UI text in different test files. Please can you have a think about whether there's a way to test this that better encapsulates things in app.ts.

Reply:
[`5ec4103`](https://github.com/microbit-foundation/python-editor-v3/pull/983/commits/5ec4103dc0c8784bab2b06d3729c48715d61466d)